src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,507 @@
+  1. **Set the minimum SDK version** back to a version before null safety.
+     ```yaml
+environment:
+    sdk: '>=2.9.0 <3.0.0'
+```
+  2. **Update dependencies:**
+     ```terminal
+$ dart pub get
+```
+
+Now, start the migration tool by running the `dart migrate` command
+in the directory that contains the package's `pubspec.yaml` file:
+
+```terminal
+$ dart migrate
+```
+
+If your package is ready to migrate,
+then the tool produces a line like the following:
+
+```terminal
+View the migration suggestions by visiting:
+
+  http://127.0.0.1:60278/Users/you/project/mypkg.console-simple?authToken=Xfz0jvpyeMI%3D
+```
+
+Visit that URL in a Chrome browser
+to see an interactive UI
+where you can guide the migration process:
+
+![Screenshot of migration tool](/null-safety/migration-tool.png)
+
+For every variable and type annotation,
+you can see what nullability the tool infers.
+For example, in the preceding screenshot,
+the tool infers that the `ints` list (previously a list of `int`)
+in line 1 is nullable, and thus should be a list of `int?`.
+
+
+#### Understanding migration results
+
+To see the reasons for each change (or non-change),
+click its line number in the **Proposed Edits** pane.
+The reasons appear in the **Edit Details** pane.
+
+For example, consider the following code,
+from before null safety:
+
+```dart
+var ints = const <int>[0, null];
+var zero = ints[0];
+var one = zero + 1;
+var zeroOne = <int>[zero, one];
+```
+
+The default migration is backward compatible but not ideal:
+
+```dart
+var ints = const <int?>[0, null];
+var zero = ints[0];
+var one = zero! + 1;
+var zeroOne = <int?>[zero, one];
+```
+
+By clicking line 3, you can see the migration tool's reasons for
+adding the `!`.
+But if you know that `zero` can't be null,
+then you can improve the migration result.
+
+
+#### Improving migration results
+
+When analysis infers the wrong nullability,
+you can override its proposed edits by inserting temporary
+`/*?*/` and `/*!*/` markers:
+
+* In the **Edit Details** pane of the migration tool,
+  you can insert these markers using the
+  **Add `/*?*/` hint** and **Add `/*!*/` hint** buttons.
+
+  These buttons add comments to your file immediately,
+  and there's no **Undo**.
+  If you don't want a hint that the tool inserted,
+  you can use your usual code editor to remove it.
+
+* You can use an editor to add these markers,
+  even while the tool is still running.
+  Because your code hasn't opted into null safety yet,
+  you can't use new null-safety features.
+  You can, however, make changes like refactoring
+  that don't depend on null-safety features.
+
+  When you've finished editing your code,
+  click **Rerun from sources** to pick up your changes.
+
+A single hint can have ripple effects elsewhere in the code.
+In the example from before,
+adding a `/*!*/` marker where `zero` is assigned its value (on line 2)
+makes the migration tool infer the type of `zero` as `int?` instead of `int`.
```

Comment:
Shouldn't this be the other way around?  `int` instead of `int?` ?

Reply:
Oops, yes.